net/rootd/src/rootd.cxx

```diff
@@ -487,7 +489,8 @@ Bool_t RootdExpandFileName(const char *fname, char *xname, const int kBufSize)
    } else if (c[0] == '~' && c[1] != '/') { // ~user case
       n = strcspn(c + 1, "/ ");
       buff[0] = 0;
-      strncat(buff, c + 1, std::min(n, kBufSize * 4));
+      std::string str(c+1);
+      strncat(buff, str.c_str(), sizeof(buff));
```

Comment:
Agreed, but `kBufSize * 4` is wrong. The length must be at most the size of the buffer being copied into, i.e. `sizeof(buff)`. The `N` here is an upper bound for the amount copied.

Reply:
> Agreed, but kBufSize * 4 is wrong

Fair enough (albeit sizeof(buff) is (currently) equal to kBufSize * 4).

> The N here is an upper bound for the amount copied.

humm .. that's not how I read the code.  It seems to be a way to chop the original string without having to modify it.

Reply:
The man page says n is for copying at most n bytes. The function is used wrongly everywhere in ROOT, by using the length of the source rather than the destination for n.

Reply:
I think you are misunderstanding me.  We need to use:
```
strncat(buff, c + 1, std::min(n, sizeof(buff)));
```
otherwise you break the function completely.

Reply:
> The function is used wrongly everywhere in ROOT, by using the length of the source rather than the destination for n.

Yes and they need to be updated to use the destination length, replacing them with a simple strcpy would open up (well keep in those case) the code exposed to memory overwrites (unless you also add the test externatlly).

Reply:
Yes, I agree with you. We need the smaller of the two. I will fix the commit.